invisible_cities/detector_simulation/buffer_functions.py

```diff
@@ -0,0 +1,201 @@
+    input Waveforms into data binned according to
+    the bin width stored in the Waveforms, effectively
+    padding with zeros inbetween the separate signals.
+
+    Parameters
+    ----------
+    max_buffer : float
+                 Maximum event time to be considered in nanoseconds
+    """
+    def bin_data(sensors  : pd.Series,
+                 bin_width: float    ,
+                 t_min    : float    ,
+                 t_max    : float    ) -> Tuple:
+        """
+        Raw data binning function.
+
+        Parameters
+        ----------
+        sensors : List of Waveforms for one sensor type
+                  Should be sorted into one type/binning
+        t_min   : float
+                  Minimum time to be used to define bins.
+        t_max   : float
+                  As t_min but the maximum to be used
+        """
+        max_time = min(t_max, t_min + max_buffer)
+        min_bin  = np.floor(t_min    / bin_width) * bin_width
+        max_bin  = np.ceil (max_time / bin_width) * bin_width
+
+        bins        = np.arange(min_bin, max_bin, bin_width)
+        b_pad       = bins[-1] + bin_width
+        bin_sensors = sensors.groupby('sensor_id').apply(weighted_histogram    ,
+                                                         np.append(bins, b_pad))
+        return bins, bin_sensors
+    return bin_data
+
+
+## !! to-do: clarify for non-pmt versions of next
+## !! to-do: Check on integral instead of only threshold?
+def signal_finder(buffer_len   : float,
+                  bin_width    : float,
+                  bin_threshold:   int) -> Callable:
+    """
+    Decides where there is signal-like
+    charge according to the configuration
+    and the PMT sum in order to give
+    a useful position for buffer selection.
+    Currently simple threshold on binned charge.
+
+    Parameters
+    ----------
+    buffer_len    : float
+                    Configured buffer length in mus
+    bin_width     : float
+                    Sampling width for sensors
+    bin_threshold : int
+                    PE threshold for selection
+    """
+
+    stand_off = int(buffer_len * units.mus / bin_width)
+    def find_signal(wfs: pd.Series) -> List[int]:
+
+        eng_sum = wfs.sum(0)
+        indices = indices_and_wf_above_threshold(eng_sum,
+                                                 bin_threshold).indices
+        ## Just using this and the stand_off for now
+        ## taking first above sum threshold.
+        ## !! To-do: make more robust with min int? or similar
+        all_indx = split_in_peaks(indices, stand_off)
+        return [pulse[0] for pulse in all_indx]
+    return find_signal
+
+
+def padder(sensors: np.ndarray, padding: Tuple) -> np.ndarray:
+    """Pads zeros around each sensor in a 2D array"""
+    if not sensors.shape[0]:
+        return np.empty((0, padding[0] + padding[1] + 1))
+    return np.apply_along_axis(np.pad, 1, sensors, padding, "constant")
```

Comment:
Padding is a Tuple because what you need to pad before the charge is rarely the same as what you need after.

If you have identified the first bin with charge as the trigger, for example, you need to pad by the number of samples corresponding to pretrigger but at the end it could be almost any number